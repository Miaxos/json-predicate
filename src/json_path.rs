use json_pointer::{JsonPointer, ParseError};
use serde::{Deserialize, Deserializer, Serialize};
use std::ops::Add;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JSONPath(JsonPointer<String, Vec<String>>);

impl Default for JSONPath {
    fn default() -> Self {
        Self(JsonPointer::new(Vec::new()))
    }
}

#[derive(Debug, thiserror::Error)]
pub enum JSONPathError {
    /// An invalid escape sequence was encountered, either a `~` escape or a
    /// `%` escape.
    #[error("An invalid escape sequence was encountered, either a `~` escape or a `%` escape.")]
    InvalidEscape,
    /// An error caused by not having a leading slash on the JSON pointer.
    ///
    /// For example, the string `a/b/c` is not a valid JSON pointer, while
    /// `/a/b/c` is.
    #[error("An error caused by not having a leading slash on the JSON pointer.")]
    NoLeadingSlash,
}

impl JSONPath {
    pub fn new<S: AsRef<str>>(value: S) -> Result<JSONPath, JSONPathError> {
        let ptr = value.as_ref().parse::<JsonPointer<_, _>>();
        match ptr {
            Ok(ptr) => Ok(JSONPath(ptr)),
            Err(ParseError::NoLeadingSlash) => Err(JSONPathError::NoLeadingSlash),
            Err(ParseError::InvalidEscape(_)) => Err(JSONPathError::InvalidEscape),
        }
    }

    pub fn empty() -> Self {
        Self::new("/").expect("Can't fail")
    }

    pub fn take(self) -> JsonPointer<String, Vec<String>> {
        self.0
    }
}

impl Serialize for JSONPath {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let val = self.0.to_string();
        Ok(serializer.serialize_str(&val)?)
    }
}

impl Add for JSONPath {
    type Output = Self;

    fn add(self, other: Self) -> Self::Output {
        // TODO: Fix it because it's an unwrap that shouldn't be here as we should be sure that it
        // works.
        Self::new(format!("{}{}", self.0.to_string(), other.0.to_string())).unwrap()
    }
}

impl<'de> Deserialize<'de> for JSONPath {
    fn deserialize<D>(deserializer: D) -> Result<JSONPath, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let ptr = s
            .parse::<JsonPointer<_, _>>()
            .map_err(|e| serde::de::Error::custom(format!("{:?}", e)))?;
        Ok(JSONPath(ptr))
    }
}
