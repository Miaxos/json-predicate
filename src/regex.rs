use std::marker::PhantomData;
use std::str::FromStr;

use regex::Regex as LibRegex;
use serde::de::{self, Deserialize, Deserializer, Visitor};
use serde::Serialize;

/// A new type wrapping a `[regex::Regex]` with:
///
/// - Eq
/// - PartialEq
/// - Serialize
/// - Deserialize
#[derive(Clone, Debug)]
pub struct Regex(LibRegex);

impl AsRef<LibRegex> for Regex {
    fn as_ref(&self) -> &LibRegex {
        &self.0
    }
}

impl From<LibRegex> for Regex {
    fn from(value: LibRegex) -> Self {
        Self(value)
    }
}

impl From<Regex> for LibRegex {
    fn from(value: Regex) -> Self {
        value.0
    }
}

impl PartialEq for Regex {
    fn eq(&self, other: &Self) -> bool {
        self.0.as_str() == other.0.as_str()
    }
}

impl Eq for Regex {}

impl Serialize for Regex {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(self.0.as_str())
    }
}

impl<'de> Deserialize<'de> for Regex {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct VisitorRegex<'de> {
            marker: PhantomData<Regex>,
            lifetime: PhantomData<&'de ()>,
        }

        impl<'de> Visitor<'de> for VisitorRegex<'de> {
            type Value = Regex;
            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("Regex")
            }

            #[inline]
            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                let inner_regex =
                    LibRegex::from_str(v).map_err(|err| serde::de::Error::custom(err))?;

                Ok(Regex(inner_regex))
            }
        }

        Deserializer::deserialize_str(
            deserializer,
            VisitorRegex {
                marker: PhantomData::<Regex>,
                lifetime: PhantomData,
            },
        )
    }
}
