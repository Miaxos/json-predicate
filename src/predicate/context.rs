use std::error::Error;

use crate::json_path::{JSONPath, JSONPathError};

/// The context to run the evaluation, you can have an empty context with
/// `PredicateContext::default()`.
#[derive(Default, Clone, Debug)]
pub struct PredicateContext {
    location: Option<JSONPath>,
}

#[derive(Debug, thiserror::Error)]
pub enum PredicateContextError {
    #[error("{0}")]
    Deserialize(#[from] Box<dyn Error>),
    #[error("{0}")]
    JSONPath(#[from] JSONPathError),
}

impl From<JSONPath> for PredicateContext {
    fn from(value: JSONPath) -> Self {
        Self {
            location: Some(value),
        }
    }
}

impl From<Option<JSONPath>> for PredicateContext {
    fn from(value: Option<JSONPath>) -> Self {
        Self { location: value }
    }
}

impl PredicateContext {
    pub fn new(path: String) -> Result<Self, PredicateContextError> {
        let location = JSONPath::new(path)?;

        Ok(Self {
            location: Some(location),
        })
    }

    pub fn final_path(&self, path: &Option<JSONPath>) -> Option<JSONPath> {
        // TODO: Remove cloning
        match (self.location.clone(), path.clone()) {
            (None, None) => None,
            (Some(path), None) | (None, Some(path)) => Some(path),
            (Some(root), Some(path)) => Some(root + path),
        }
    }
}
