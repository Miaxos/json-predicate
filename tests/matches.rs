// Test adapted from https://github.com/MalcolmDwyer/json-predicate/blob/master/test/test.js

use json_predicate::builder::MatchesBuilder;
use json_predicate::context::PredicateContext;
use json_predicate::json_path::JSONPath;
use json_predicate::{FirstOrder, Predicate, PredicateImpl};

mod utils;
use regex::Regex;
use serde::Deserialize;
use serde_json::Value;
use utils::ENTRY;

#[test]
pub fn test_base_predicate_snapshot() {
    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/stringABC").unwrap())
            .ignore_case(false)
            .value(Regex::new(".*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    insta::assert_json_snapshot!(predicate);
}

#[test]
pub fn returns_false_for_any_non_string_target_path() {
    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/num1".to_owned()).unwrap())
            .value(Regex::new("[\\w\\s]*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);

    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/null1".to_owned()).unwrap())
            .value(Regex::new("[\\w\\s]*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);

    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/objA".to_owned()).unwrap())
            .value(Regex::new("[\\w\\s]*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);

    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/arrayA".to_owned()).unwrap())
            .value(Regex::new("[\\w\\s]*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);
}

#[test]
pub fn returns_false_for_string_that_would_make_an_invalid_regex() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": "\\",
    }));

    assert!(predicate.is_err());
    insta::assert_debug_snapshot!(predicate.unwrap_err(), @r###"Error("data did not match any variant of untagged enum Predicate", line: 0, column: 0)"###);
}

#[test]
pub fn returns_false_for_any_value_other_than_a_regexp_or_a_string() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": 1,
    }));

    assert!(predicate.is_err());

    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": { "a": 1},
    }));

    assert!(predicate.is_err());

    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": ["a", "b"],
    }));

    assert!(predicate.is_err());
}

#[test]
pub fn returns_true_for_match_when_providing_string_that_will_become_regexp() {
    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/stringABC".to_owned()).unwrap())
            .value(Regex::new("[A-Z]*").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), true);
}

#[test]
pub fn returns_true_for_different_case_string_with_ignore_case() {
    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/stringABC".to_owned()).unwrap())
            .value(Regex::new("aBc").unwrap())
            .ignore_case(true)
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), true);
}

#[test]
pub fn returns_false_for_different_case_string_without_ignore_case_true() {
    let predicate: Predicate = FirstOrder::from(
        MatchesBuilder::default()
            .path(JSONPath::new("/stringABC".to_owned()).unwrap())
            .value(Regex::new("aBc").unwrap())
            .build()
            .unwrap(),
    )
    .into();

    let result = predicate.evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);
}

#[test]
#[ignore = "failing"]
pub fn returns_true_when_providing_matching_regexp_directly() {
    let json = r###"
    {
      "op": "matches",
      "path": "/stringABC",
      "value": /[A-Z]+/
    }
    "###;
    let predicate = Predicate::deserialize(serde_json::from_str::<Value>(json).unwrap());

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), true);
}

#[test]
pub fn returns_false_when_providing_matching_regex_with_mismatch_case() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": "/[a-z]+",
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), false);
}

#[test]
pub fn returns_true_when_providing_matching_i_regexp_without_mismatch_case() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/stringABC",
        "value": "(?i)[a-z]+",
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), true);
}

#[test]
pub fn returns_true_when_providing_matching_i_regexp_with_mismatch_case() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches-",
        "path": "/stringABC",
        "value": "(?i)[a-z]+",
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), true);
}

#[test]
pub fn returns_err_for_undefined_value() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches",
        "path": "/objZZZ/objZZZZZZZZ",
        "value": "[whatever]",
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_err());
}

#[test]
pub fn returns_err_for_undefined_value_with_ignore_case() {
    let predicate = Predicate::deserialize(serde_json::json!({
        "op": "matches-",
        "path": "/objZZZ/objZZZZZZZZ",
        "value": "[whatever]",
    }));

    assert!(predicate.is_ok());
    let result = predicate
        .unwrap()
        .evaluate(&ENTRY, PredicateContext::default());
    assert!(result.is_err());
}
