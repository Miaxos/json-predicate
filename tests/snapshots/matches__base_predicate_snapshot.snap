---
source: tests/matches.rs
expression: predicate

---
{
  "op": "matches",
  "path": "/stringABC",
  "value": ".*"
}
